### PR TITLE
Some fixes / improvements for AMP

### PR DESCRIPTION
**Pull request by nickwp**
Fix AMP execution flow, scheduler stepping, and scaler checkpointing

- Initialize GradScaler only when AMP is explicitly enabled in `configure_amp` instead of creating it unconditionally on CUDA
- Run `forward_pass()` + `compute_metrics()` under `autocast` in `step()` so AMP is used consistently for training, validation and inference, not just for training
- Only do `scheduler.step()` when the optimizer actually updated under AMP by ensuring the scale didn't decrease, to avoid LR advancing on skipped overflow steps that cause the scale to decrease
- Save and restore AMP scaler state in checkpoints

### CODE REVIEW

**Comment by nickwp** [comment]
@cshuoyu  I have made some changes to your AMP implementation. Could you check that these changes look good to you?

**Comment by cshuoyu** [comment]
Hi, I've read these changes and tested. They looks good for me. Thank you very much.

**Review by cshuoyu — APPROVED** [review]

**Comment by nickwp** [comment]
Thanks @cshuoyu !
